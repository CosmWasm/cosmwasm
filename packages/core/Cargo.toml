[package]
name = "cosmwasm-core"
version = { workspace = true }
edition = "2021"
description = "Components of cosmwasm-std that can be used in no_std environments"
repository = "https://github.com/CosmWasm/cosmwasm/tree/main/packages/core"
license = "Apache-2.0"
readme = "README.md"
