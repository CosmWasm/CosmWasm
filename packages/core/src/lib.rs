//! cosmwasm-core contains components of cosmwasm-std that can be used in a no_std environment.
//! All symbols are re-exported by cosmwasm-std, such that contract developers don't need to
//! add this dependency directly. It is recommended to only use cosmwasm-std whenever possible.

#![cfg_attr(not(feature = "std"), no_std)]

#[macro_use]
extern crate alloc;

#[cfg(test)]
extern crate std; // allow for file I/O during tests

mod addresses;
mod binary;
mod encoding;
mod errors;
mod forward_ref;
mod hex_binary;
mod math;
mod timestamp;

#[doc(hidden)]
pub mod __internal;

use crate::__internal__forward_ref_partial_eq as forward_ref_partial_eq;

pub use crate::addresses::{instantiate2_address, Addr, CanonicalAddr, Instantiate2AddressError};
pub use crate::binary::Binary;
pub use crate::encoding::{from_base64, from_hex, to_base64, to_hex};
pub use crate::errors::{
    AggregationError, AggregationPairingEqualityError, CheckedFromRatioError,
    CheckedMultiplyFractionError, CheckedMultiplyRatioError, CoinFromStrError, CoinsError,
    ConversionOverflowError, CoreError, CoreResult, DivideByZeroError, DivisionError,
    OverflowError, OverflowOperation, RecoverPubkeyError, RoundDownOverflowError,
    RoundUpOverflowError, SystemError, VerificationError,
};
pub use crate::hex_binary::HexBinary;
pub use crate::math::{
    Decimal, Decimal256, Decimal256RangeExceeded, DecimalRangeExceeded, Fraction, Int128, Int256,
    Int512, Int64, Isqrt, SignedDecimal, SignedDecimal256, SignedDecimal256RangeExceeded,
    SignedDecimalRangeExceeded, Uint128, Uint256, Uint512, Uint64,
};
pub use crate::timestamp::Timestamp;

/// Exposed for testing only
/// Both unit tests and integration tests are compiled to native code, so everything in here does not need to compile to Wasm.
#[cfg(not(target_arch = "wasm32"))]
pub mod testing;
