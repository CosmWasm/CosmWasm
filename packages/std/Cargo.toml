[package]
name = "cosmwasm-std"
version = "1.2.5"
authors = ["Ethan Frey <ethanfrey@users.noreply.github.com>"]
edition = "2021"
description = "Standard library for Wasm based smart contracts on Cosmos blockchains"
repository = "https://github.com/CosmWasm/cosmwasm/tree/main/packages/std"
license = "Apache-2.0"
readme = "README.md"

[package.metadata.docs.rs]
features = ["abort", "stargate", "staking", "ibc3", "cosmwasm_1_2"]

[features]
default = ["iterator", "abort", "std"]
std = ["base64/std", "uint/std", "serde-json-wasm/std", "forward_ref", "schemars", "thiserror/std"]
abort = []
# iterator allows us to iterate over all DB items in a given range
# optional as some merkle stores (like tries) don't support this
# given Ethereum 1.0, 2.0, Substrate, and other major projects use Tries
# we keep this optional, to allow possible future integration (or different Cosmos Backends)
iterator = []
# staking exposes bindings to a required staking moudle in the runtime, via new
# CosmosMsg types, and new QueryRequest types. This should only be enabled on contracts
# that require these types, so other contracts can be used on systems with eg. PoA consensus
staking = []
# backtraces provides much better context at runtime errors (in non-wasm code)
# at the cost of a bit of code size and performance.
# This feature requires Rust nightly because it depends on the unstable backtrace feature.
backtraces = []
# stargate enables stargate-dependent messages and queries, like raw protobuf messages
# as well as ibc-related functionality
stargate = []
# ibc3 extends ibc messages with ibc-v3 only features. This should only be enabled on contracts
# that require these types. Without this, they get the smaller ibc-v1 API.
ibc3 = ["stargate"]
# This feature makes `BankQuery::Supply` available for the contract to call, but requires
# the host blockchain to run CosmWasm `1.1.0` or higher.
cosmwasm_1_1 = []
# This feature makes `GovMsg::VoteWeighted` available for the contract to call, but requires
# the host blockchain to run CosmWasm `1.2.0` or higher.
cosmwasm_1_2 = ["cosmwasm_1_1"]


[dependencies]
base64 =  {version = "0.13.0", default-features = false, features = ["alloc"]}
cosmwasm-derive = { path = "../derive", version = "1.2.5" }
derivative = { version = "2", default-features = false, features = ["use_core"] }
hex = { version = "0.4", default-features = false, features = ["alloc"] }
schemars = {version = "0.8.3", optional = true}
sha2 = { version = "0.10.3", default-features = false}
serde = { version = "1.0.103", default-features = false, features = ["derive", "alloc"] }
serde-json-wasm = { git = "https://github.com/dzmitry-lahoda-forks/serde-json-wasm", rev = "8a7e522c0e4de36a6dfb535766f26a9941017d81", default-features = false}
thiserror = { version = "1.0.38", package = "thiserror-core", default-features = false }

uint = { version = "0.9.3", default-features = false }

forward_ref = { version = "1", default-features = false, optional = true}

[target.'cfg(not(target_arch = "wasm32"))'.dependencies]
cosmwasm-crypto = { path = "../crypto", version = "1.2.5", optional = true}


[dev-dependencies]
cosmwasm-schema = { path = "../schema" }
# The chrono dependency is only used in an example, which Rust compiles for us. If this causes trouble, remove it.
chrono = { version = "0.4", default-features = false, features = ["alloc", "std"] }
hex-literal = "0.3.1"
serde_json = "1.0.81"
