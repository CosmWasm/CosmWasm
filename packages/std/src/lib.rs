#[macro_use]
extern crate alloc;

// Exposed on all platforms

mod assertions;
mod checksum;
mod coin;
mod coins;
mod conversion;
mod deps;
mod ibc;
mod import_helpers;
#[cfg(feature = "iterator")]
mod iterator;
mod metadata;
mod never;
mod pagination;
mod panic;
mod query;
mod results;
mod sections;
mod serde;
mod stdack;
mod storage;
mod traits;
mod types;
mod uuid;

/// This module is to simplify no_std imports
pub(crate) mod prelude;

/// This modules is very advanced and will not be used directly by the vast majority of users.
/// We want to offer it to ensure a stable storage key composition system but don't encourage
/// contract devs to use it directly.
pub mod storage_keys;

pub use crate::checksum::{Checksum, ChecksumError};
pub use crate::coin::{coin, coins, has_coins, Coin};
pub use crate::coins::Coins;
pub use crate::deps::{Deps, DepsMut, OwnedDeps};
pub use crate::ibc::IbcChannelOpenResponse;
pub use crate::ibc::{
    Ibc3ChannelOpenResponse, IbcAcknowledgement, IbcBasicResponse, IbcChannel, IbcChannelCloseMsg,
    IbcChannelConnectMsg, IbcChannelOpenMsg, IbcEndpoint, IbcMsg, IbcOrder, IbcPacket,
    IbcPacketAckMsg, IbcPacketReceiveMsg, IbcPacketTimeoutMsg, IbcReceiveResponse, IbcTimeout,
    IbcTimeoutBlock,
};
#[cfg(feature = "iterator")]
pub use crate::iterator::{Order, Record};
pub use crate::metadata::{DenomMetadata, DenomUnit};
pub use crate::never::Never;
pub use crate::pagination::PageRequest;
pub use crate::query::{
    AllBalanceResponse, AllDelegationsResponse, AllDenomMetadataResponse, AllValidatorsResponse,
    BalanceResponse, BankQuery, BondedDenomResponse, ChannelResponse, CodeInfoResponse,
    ContractInfoResponse, CustomQuery, DecCoin, Delegation, DelegationResponse,
    DelegationRewardsResponse, DelegationTotalRewardsResponse, DelegatorReward,
    DelegatorValidatorsResponse, DelegatorWithdrawAddressResponse, DenomMetadataResponse,
    DistributionQuery, FullDelegation, GrpcQuery, IbcQuery, ListChannelsResponse, PortIdResponse,
    QueryRequest, StakingQuery, SupplyResponse, Validator, ValidatorResponse, WasmQuery,
};
#[cfg(all(feature = "stargate", feature = "cosmwasm_1_2"))]
pub use crate::results::WeightedVoteOption;
pub use crate::results::{
    attr, wasm_execute, wasm_instantiate, AnyMsg, Attribute, BankMsg, ContractResult, CosmosMsg,
    CustomMsg, Empty, Event, MsgResponse, QueryResponse, Reply, ReplyOn, Response, SubMsg,
    SubMsgResponse, SubMsgResult, SystemResult, WasmMsg,
};
#[cfg(feature = "staking")]
pub use crate::results::{DistributionMsg, StakingMsg};
#[cfg(feature = "stargate")]
pub use crate::results::{GovMsg, VoteOption};
#[allow(deprecated)]
pub use crate::serde::{
    from_binary, from_json, from_slice, to_binary, to_json_binary, to_json_string, to_json_vec,
    to_vec,
};
pub use crate::stdack::StdAck;
pub use crate::storage::MemoryStorage;
pub use crate::traits::{Api, Querier, QuerierResult, QuerierWrapper, Storage};
pub use crate::types::{BlockInfo, ContractInfo, Env, MessageInfo, TransactionInfo};
pub use crate::uuid::{new_uuid, Uuid};

// Exposed in wasm build only

#[cfg(target_arch = "wasm32")]
mod exports;
#[cfg(target_arch = "wasm32")]
mod imports;
#[cfg(target_arch = "wasm32")]
mod memory; // Used by exports and imports only. This assumes pointers are 32 bit long, which makes it untestable on dev machines.

#[cfg(target_arch = "wasm32")]
pub use crate::exports::{do_execute, do_instantiate, do_migrate, do_query, do_reply, do_sudo};
#[cfg(all(feature = "stargate", target_arch = "wasm32"))]
pub use crate::exports::{
    do_ibc_channel_close, do_ibc_channel_connect, do_ibc_channel_open, do_ibc_packet_ack,
    do_ibc_packet_receive, do_ibc_packet_timeout,
};
#[cfg(target_arch = "wasm32")]
pub use crate::imports::{ExternalApi, ExternalQuerier, ExternalStorage};

/// Exposed for testing only
/// Both unit tests and integration tests are compiled to native code, so everything in here does not need to compile to Wasm.
#[cfg(not(target_arch = "wasm32"))]
pub mod testing;

// Re-exports

pub use cosmwasm_core::CoreError as StdError;
pub use cosmwasm_core::CoreResult as StdResult;
pub use cosmwasm_core::{
    from_base64, from_hex, instantiate2_address, to_base64, to_hex, Addr, Binary, CanonicalAddr,
    CheckedFromRatioError, CheckedMultiplyFractionError, CheckedMultiplyRatioError,
    CoinFromStrError, CoinsError, ConversionOverflowError, Decimal, Decimal256,
    Decimal256RangeExceeded, DecimalRangeExceeded, DivideByZeroError, DivisionError, Fraction,
    HexBinary, Instantiate2AddressError, Int128, Int256, Int512, Int64, Isqrt, OverflowError,
    OverflowOperation, RecoverPubkeyError, SignedDecimal, SignedDecimal256,
    SignedDecimal256RangeExceeded, SignedDecimalRangeExceeded, SystemError, Timestamp, Uint128,
    Uint256, Uint512, Uint64, VerificationError,
};

#[cfg(not(target_arch = "wasm32"))]
pub use cosmwasm_core::assert_approx_eq;

pub use cosmwasm_derive::entry_point;
