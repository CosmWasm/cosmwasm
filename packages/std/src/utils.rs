/// Implements a hidden constructor for non-exhaustive structures that need
/// to be built in libraries like cw-multi-test.
macro_rules! impl_hidden_constructor {
    ( $response:ty, $( $field: ident : $t: ty),* ) => {
        impl $response {
            /// Constructor for testing frameworks such as cw-multi-test.
            /// This is required because the type is #[non_exhaustive].
            /// As a contract developer you should not need this constructor since
            /// the given structure is constructed for you via deserialization.
            ///
            /// Warning: This can change in breaking ways in minor versions.
            #[doc(hidden)]
            #[allow(dead_code)]
            pub fn new($( $field: $t),*) -> Self {
                Self { $( $field ),* }
            }
        }
    };
}

pub(crate) use impl_hidden_constructor;
