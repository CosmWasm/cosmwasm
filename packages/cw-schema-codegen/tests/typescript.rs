use core::str;
use std::{fs::File, io::Write, process::Command};

use crate::utils::TestCase;

mod utils;

#[test]
fn codegen_snap() {
    // generate the schemas for each of the above types
    let schemas = [
        cw_schema::schema_of::<utils::Owo>(),
        cw_schema::schema_of::<utils::Uwu>(),
        cw_schema::schema_of::<utils::Òwó>(),
        cw_schema::schema_of::<utils::Empty>(),
        cw_schema::schema_of::<utils::Hehehe>(),
    ];

    // run the codegen to typescript
    for schema in schemas {
        let cw_schema::Schema::V1(schema) = schema else {
            panic!();
        };

        let output = schema
            .definitions
            .iter()
            .map(|node| {
                let mut buf = Vec::new();
                cw_schema_codegen::typescript::process_node(&mut buf, &schema, node, true).unwrap();
                String::from_utf8(buf).unwrap()
            })
            .collect::<String>();

        insta::assert_snapshot!(output);
    }
}

#[test]
#[ignore] // because it requires NPM to be installed, CI will still run it
fn assert_validity() {
    let e2e_dir = format!("{}/tests/ts-e2e", env!("CARGO_MANIFEST_DIR"));
    let gen_file_path = format!("{}/src/gen.ts", e2e_dir);

    // make sure the dependencies are installed
    let install_status = Command::new("npm")
        .arg("i")
        .current_dir(&e2e_dir)
        .status()
        .unwrap();
    assert!(install_status.success());

    utils::run_e2e(
        |buf, schema, node| cw_schema_codegen::typescript::process_node(buf, schema, node, true),
        |TestCase { code, type_name }| {
            let mut gen_file = File::create(&gen_file_path).unwrap();
            gen_file.write_all(code.as_bytes()).unwrap();

            let mut cmd = Command::new("npm");
            cmd.args(["test".into(), format!("{type_name}Schema")])
                .current_dir(&e2e_dir);

            cmd
        },
    );
}
