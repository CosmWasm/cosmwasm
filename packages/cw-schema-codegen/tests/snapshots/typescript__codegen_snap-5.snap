---
source: packages/cw-schema-codegen/tests/typescript.rs
expression: output
---
// This code is @generated by cw-schema-codegen. Do not modify this manually.


import { z } from 'zod';


/**

 */

const typescript_utils_HeheheSchema = z.union([

    /**
    
     */


    z.object({ "A": z.null() }).or(z.literal("A")),


    /**
    
     */


    z.object({
      "B": z.tuple([z.string().or(z.number())])
      
        .or(z.string().or(z.number()))
      
  })

    ,


    /**
    
     */


    z.object({ "C": z.object({

        /**
        
         */

        field: z.string(),

}) }),




]);

type typescript_utils_Hehehe = z.infer<typeof typescript_utils_HeheheSchema>;

export { typescript_utils_Hehehe, typescript_utils_HeheheSchema };
